Make thread pools more robust

milljm opened the pull request:
The locking mechanisms needed some work when determining
whether or not to submit a job to the thread_pools.

Added more keyboard exceptions to beautify ctrl-c exit
messages.

Closes #10779

Commit:
Make thread pools more robust

The locking mechanisms needed some work when determining
whether or not to submit a job to the thread_pools.

Added more keyboard exceptions to beautify ctrl-c exit
messages.

Closes #10779

moosebuild commented:
Job [Documentation](https://www.moosebuild.org/job/157131/) on 96e04a0 wanted to post the following:

View the site [here](http://mooseframework.org/docs/PRs/10787/site)

This comment will be updated on new commits.

brianmoose commented on python/TestHarness/schedulers/Scheduler.py:
```diff
@@ -339,60 +350,59 @@ def schedule(self, testers):
+            while self.job_queue_count > 0:
+                # One of our children died :( so exit uncleanly (do not join)
+                if self.error_state:
+                    return
```
What happens to currently running threads? Should this do a `killRemaining()` before returning?

Commit:
Attempt to kill remaining jobs

When we hit a scheduler error or interrupt, attempt
to kill remaining jobs instead of just returning.

Closes #10779
